tools/tsp-client/src/index.ts

```diff
@@ -1,62 +1,217 @@
+            Logger.debug(`Service directory: ${serviceDir}`)
+            mkdir(path.join(outputDir, serviceDir), { recursive: true }).then(() => {
+                writeFile(path.join(path.join(outputDir, serviceDir), "tsp-location.yaml"), `directory: ${matchRes.groups!["path"]!}\ncommit: ${matchRes.groups!["commit"]!}\nrepo: ${matchRes.groups!["repo"]!}`);
+            });
+            if (configYaml["parameters"]["dependencies"] && configYaml["parameters"]["dependencies"]["additionalDirectories"]) {
+              const additionalDirs = configYaml["parameters"]["dependencies"]["additionalDirectories"];
+              // FIXME add this
+              Logger.info(`Additional directories: ${additionalDirs}`)
+            }
+            return path.join(outputDir, serviceDir);
+          } else {
+            Logger.error("Missing service-dir in parameters section of tspconfig.yaml. Please refer to https://github.com/Azure/azure-rest-api-specs/blob/main/specification/contosowidgetmanager/Contoso.WidgetManager/tspconfig.yaml for the right schema.")
+          }
+        }
+      }
+    }
+  } else {
+    // File scenario
+    throw new Error("File scenario not implemented yet")
   }
-  Logger.debug(`Using base url from  tsp-location.yaml: ${tspUrl}`);
-  for (const mainFile of ["client.tsp", "main.tsp"]) {
-    let url = new URL(mainFile, tspUrl).toString();
-    Logger.debug(`Checking for file: ${url}`);
-    if (await doesFileExist(url)) {
-      Logger.debug(`Found main file: ${url}`);
-      return url;
+  throw new Error("Invalid tspconfig.yaml");  
+}
+
+async function syncTspFiles(outputDir: string) {
+  const tempRoot = await createTempDirectory(outputDir);
+
+  const repoRoot = getRepoRoot();
+  Logger.debug(`Repo root is ${repoRoot}`);
+  if (repoRoot === undefined) {
+    throw new Error("Could not find repo root");
+  }
+
+  const cloneDir = path.join(repoRoot, "..", "sparse-spec");
+  Logger.debug(`Cloning repo to ${cloneDir}`);
+  const [ directory, commit, repo, additionalDirectories ] = await readTspLocation(outputDir);
+  const dirSplit = directory.split("/");
+  var projectName = dirSplit[dirSplit.length - 1];
+  Logger.debug(`Using project name: ${projectName}`)
+  if (projectName === undefined) {
+    projectName = "src";
+  }
+  const srcDir = path.join(tempRoot, projectName);
+  mkdir(srcDir, { recursive: true });
+  if (existsSync(cloneDir)) {
+    Logger.debug(`Removing existing sparse-checkout directory ${cloneDir}`);
+    await removeDirectory(cloneDir);
+  }
+  await cloneRepo(tempRoot, cloneDir, `https://github.com/${repo}.git`);
+  await sparseCheckout(cloneDir);
+  await addSpecFiles(cloneDir, directory)
+  Logger.info(`Processing additional directories: ${additionalDirectories}`)
+  for (const dir of additionalDirectories) {
+    await addSpecFiles(cloneDir, dir);
+  }
+  await checkoutCommit(cloneDir, commit);
+  
+  cp(path.join(cloneDir, directory), srcDir, { recursive: true }, (err) => {
```

Comment:
can you use the promisified version of cp instead of the old callback version? This won't do what you think it will.